File: reference/scanner/oracle_risk.py
"""Analyse de fiabilité des oracles Morpho (la leçon Resolv).

Le vrai péché du hack Resolv : un oracle qui valorisait le collatéral à un peg
hardcodé / une NAV contrôlée par l'émetteur, alors que le prix marché s'effondrait.
On détecte programmatiquement ces patterns depuis `oracle.data` + `warnings`.
"""
from __future__ import annotations

ZERO = "0x0000000000000000000000000000000000000000"

# Hardcoder le peg d'un stable majeur (USDC=$1) est standard sur Morpho et peu
# risqué. Le danger Resolv venait du hardcode/NAV sur un collatéral exotique.
MAJOR_STABLES = {"usdc", "usdt", "dai", "usds", "frax", "ausd"}


def _addr(feed: dict | None) -> str | None:
    if not feed:
        return None
    a = feed.get("address")
    return None if (a is None or a == ZERO) else a


def analyze_oracle(market: dict) -> dict:
    """Renvoie les flags de risque oracle d'un marché."""
    o = market.get("oracle") or {}
    data = o.get("data") or {}
    flags: list[str] = []
    otype = o.get("type")

    # 1. oracle opaque / non reconnu
    if otype in (None, "Unknown", "CustomOracle") or data.get("__typename") is None:
        flags.append("opaque_oracle")

    # 2. peg hardcodé : une jambe de feed à 0x0 => prix de cette jambe = 1.
    #    Côté quote (loan) nul avec un stable => "assume <loan> = $1".
    if data.get("__typename") == "MorphoChainlinkOracleV2Data":
        qf1 = _addr(data.get("quoteFeedOne"))
        bf1 = _addr(data.get("baseFeedOne"))
        loan = (market.get("loanAsset") or {}).get("symbol", "?")
        col = (market.get("collateralAsset") or {}).get("symbol", "?")
        if qf1 is None and not _vault(data.get("quoteOracleVault")):
            kind = "peg_assumption" if loan.lower() in MAJOR_STABLES else "hardcoded_peg"
            flags.append(f"{kind}(assume {loan}=$1)")
        if bf1 is None and not _vault(data.get("baseOracleVault")):
            kind = "peg_assumption" if col.lower() in MAJOR_STABLES else "hardcoded_peg"
            flags.append(f"{kind}(assume {col}=$1)")

        # 3. dépendance NAV / exchange-rate (ERC-4626 mis à jour par l'émetteur)
        if _vault(data.get("baseOracleVault")):
            flags.append("nav_dependency(collateral)")
        if _vault(data.get("quoteOracleVault")):
            flags.append("nav_dependency(loan)")

    # 4. warnings émis directement par Morpho (les plus forts)
    for w in (market.get("warnings") or []):
        t = w.get("type")
        lvl = w.get("level")
        if t == "oracle_price_derivation":
            flags.append(f"PRICE_DERIVATION_DIVERGENCE[{lvl}]")
        elif t in ("bad_debt_unrealized", "bad_debt_realized"):
            flags.append(f"{t.upper()}[{lvl}]")
        elif t in ("unrecognized_collateral_asset", "unrecognized_loan_asset"):
            flags.append(t)

    return {
        "oracle_type": otype,
        "oracle_address": o.get("address"),
        "flags": flags,
        "severity": _severity(flags),
    }


def _vault(v: dict | None) -> bool:
    return bool(v and v.get("address") and v["address"] != ZERO)


def _severity(flags: list[str]) -> str:
    blob = " ".join(flags).lower()
    if "price_derivation" in blob or "bad_debt" in blob or "opaque" in blob:
        return "RED"
    if "hardcoded_peg" in blob or "nav_dependency" in blob:
        return "YELLOW"
    return "OK"


def contagion_buckets(positions: list[dict]) -> dict:
    """Regroupe les positions par hypothèse oracle partagée.

    positions: [{label, usd, oracle: <analyze_oracle result>}]
    -> 'si telle hypothèse casse, $X exposés sur N marchés'.
    """
    buckets: dict[str, dict] = {}
    for p in positions:
        for f in p["oracle"]["flags"]:
            key = f.split("[")[0]  # normalise les niveaux
            b = buckets.setdefault(key, {"usd": 0.0, "markets": 0, "examples": []})
            b["usd"] += p["usd"]
            b["markets"] += 1
            if len(b["examples"]) < 4:
                b["examples"].append(p["label"])
    return dict(sorted(buckets.items(), key=lambda kv: -kv[1]["usd"]))

File: lib/oracleRisk.ts
// Analyse de fiabilité des oracles Morpho (la leçon Resolv).
// Port TS de reference/scanner/oracle_risk.py.

import type { Market, OracleVaultRef, OracleFeed, OracleAnalysis, Position, ContagionBucket } from "./types";

const ZERO = "0x0000000000000000000000000000000000000000";
// Hardcoder le peg d'un stable majeur (USDC=$1) est standard et peu risqué.
const MAJOR_STABLES = new Set(["usdc", "usdt", "dai", "usds", "frax", "ausd"]);

function feedAddr(feed: OracleFeed): string | null {
  if (!feed || !feed.address || feed.address === ZERO) return null;
  return feed.address;
}
function hasVault(v: OracleVaultRef): boolean {
  return !!(v && v.address && v.address !== ZERO);
}

export function analyzeOracle(market: Market): OracleAnalysis {
  const o = market.oracle ?? null;
  const data = o?.data ?? null;
  const flags: string[] = [];
  const otype = o?.type ?? null;

  // Opaque UNIQUEMENT si aucune donnée d'oracle exploitable. NB: l'API renvoie
  // souvent type="Unknown" alors que oracle.data est un MorphoChainlinkOracleV2Data
  // valide (feeds Chainlink présents) -> ne PAS le flagger opaque dans ce cas.
  if (!data?.__typename) {
    flags.push("opaque_oracle");
  }

  if (data?.__typename === "MorphoChainlinkOracleV2Data") {
    const loan = market.loanAsset?.symbol ?? "?";
    const col = market.collateralAsset?.symbol ?? "?";
    if (feedAddr(data.quoteFeedOne ?? null) === null && !hasVault(data.quoteOracleVault ?? null)) {
      const kind = MAJOR_STABLES.has(loan.toLowerCase()) ? "peg_assumption" : "hardcoded_peg";
      flags.push(`${kind}(assume ${loan}=$1)`);
    }
    if (feedAddr(data.baseFeedOne ?? null) === null && !hasVault(data.baseOracleVault ?? null)) {
      const kind = MAJOR_STABLES.has(col.toLowerCase()) ? "peg_assumption" : "hardcoded_peg";
      flags.push(`${kind}(assume ${col}=$1)`);
    }
    if (hasVault(data.baseOracleVault ?? null)) flags.push("nav_dependency(collateral)");
    if (hasVault(data.quoteOracleVault ?? null)) flags.push("nav_dependency(loan)");
  }

  for (const w of market.warnings ?? []) {
    if (w.type === "oracle_price_derivation") flags.push(`PRICE_DERIVATION_DIVERGENCE[${w.level}]`);
    else if (w.type === "bad_debt_unrealized" || w.type === "bad_debt_realized") flags.push(`${w.type.toUpperCase()}[${w.level}]`);
    else if (w.type === "unrecognized_collateral_asset" || w.type === "unrecognized_loan_asset") flags.push(w.type);
  }

  return { oracle_type: otype, oracle_address: o?.address ?? null, flags, severity: severity(flags) };
}

function severity(flags: string[]): "RED" | "YELLOW" | "OK" {
  const blob = flags.join(" ").toLowerCase();
  if (blob.includes("price_derivation") || blob.includes("bad_debt") || blob.includes("opaque")) return "RED";
  if (blob.includes("hardcoded_peg") || blob.includes("nav_dependency")) return "YELLOW";
  return "OK";
}

export function contagionBuckets(positions: Position[]): Record<string, ContagionBucket> {
  const buckets: Record<string, ContagionBucket> = {};
  for (const p of positions) {
    for (const f of p.oracle.flags) {
      const key = f.split("[")[0];
      const b = (buckets[key] ??= { usd: 0, markets: 0, examples: [] });
      b.usd += p.usd;
      b.markets += 1;
      if (b.examples.length < 4) b.examples.push(p.label);
    }
  }
  return Object.fromEntries(Object.entries(buckets).sort((a, b) => b[1].usd - a[1].usd));
}
